subsys/usb/device_next/usbd_endpoint.c

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/usb/udc.h>
+#include <zephyr/usb/usbd.h>
+
+#include "usbd_device.h"
+#include "usbd_class.h"
+#include "usbd_ch9.h"
+#include "usbd_desc.h"
+#include "usbd_endpoint.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(usbd_ep, CONFIG_USBD_LOG_LEVEL);
+
+int usbd_ep_enable(const struct device *dev,
+		   const struct usb_ep_descriptor *const ed,
+		   uint32_t *const ep_bm)
+{
+	int ret;
+
+	ret = udc_ep_enable(dev, ed->bEndpointAddress, ed->bmAttributes,
+			    ed->wMaxPacketSize, ed->bInterval);
+	if (ret == 0) {
+		usbd_ep_bm_set(ep_bm, ed->bEndpointAddress);
+	}
+
+	return ret;
+}
+
+int usbd_ep_disable(const struct device *dev,
+		    const uint8_t ep,
+		    uint32_t *const ep_bm)
+{
+	int ret;
+
+	ret = udc_ep_disable(dev, ep);
+	if (ret) {
+		return ret;
+	}
+
+	usbd_ep_bm_clear(ep_bm, ep);
+
+	ret = udc_ep_dequeue(dev, ep);
```

Comment:
This should be called only if the EP was enqueued. Unconditionally dequeueing the EP leads to `<err> udc_nrf: ep 0x01 queue is empty` after disconnecting the device running CDC ACM sample.

Reply:
If the check is supposed to happen in the driver, then the `udc_ep_dequeue()` API documentation should mention it. Otherwise, there'll be some drivers that do handle it and some that don't.